src/main/java/com/linecorp/armeria/server/AbstractVirtualHostBuilder.java

```diff
@@ -0,0 +1,277 @@
+            if (line == null) {
+                logger.warn("The 'hostname' command returned nothing; " +
+                            "using InetAddress.getLocalHost() instead", line);
+            } else {
+                hostname = normalizeDefaultHostname(line.trim());
+                logger.info("Hostname: {} (via 'hostname' command)", hostname);
+            }
+        } catch (Exception e) {
+            logger.warn("Failed to get the hostname using the 'hostname' command; " +
+                        "using InetAddress.getLocalHost() instead", e);
+        } finally {
+            if (process != null) {
+                process.destroy();
+            }
+        }
+
+        if (hostname == null) {
+            try {
+                hostname = normalizeDefaultHostname(InetAddress.getLocalHost().getHostName());
+                logger.info("Hostname: {} (via InetAddress.getLocalHost())", hostname);
+            } catch (Exception e) {
+                hostname = "localhost";
+                logger.warn("Failed to get the hostname using InetAddress.getLocalHost(); " +
+                            "using 'localhost' instead", e);
+            }
+        }
+
+        LOCAL_HOSTNAME = hostname;
+    }
+
+    private final String defaultHostname;
+    private final String hostnamePattern;
+    private final List<ServiceConfig> services = new ArrayList<>();
+    private SslContext sslContext;
+    private Function<Service<Request, Response>, Service<Request, Response>> decorator;
+
+    /**
+     * Creates a new {@link VirtualHostBuilder} whose hostname pattern is {@code "*"} (match-all).
+     */
+    AbstractVirtualHostBuilder() {
+        this(LOCAL_HOSTNAME, "*");
+    }
+
+    /**
+     * Creates a new {@link VirtualHostBuilder} with the specified hostname pattern.
+     */
+    AbstractVirtualHostBuilder(String hostnamePattern) {
+        hostnamePattern = normalizeHostnamePattern(hostnamePattern);
+
+        if ("*".equals(hostnamePattern)) {
+            defaultHostname = LOCAL_HOSTNAME;
+        } else if (hostnamePattern.startsWith("*.")) {
+            defaultHostname = hostnamePattern.substring(2);
+        } else {
+            defaultHostname = hostnamePattern;
+        }
+
+        this.hostnamePattern = hostnamePattern;
+    }
+
+    /**
+     * Creates a new {@link VirtualHostBuilder} with the specified hostname pattern.
+     */
+    AbstractVirtualHostBuilder(String defaultHostname, String hostnamePattern) {
+        requireNonNull(defaultHostname, "defaultHostname");
+
+        defaultHostname = normalizeDefaultHostname(defaultHostname);
+        hostnamePattern = normalizeHostnamePattern(hostnamePattern);
+        ensureHostnamePatternMatchesDefaultHostname(hostnamePattern, defaultHostname);
+
+        this.defaultHostname = defaultHostname;
+        this.hostnamePattern = hostnamePattern;
+    }
+
+    /**
+     * Sets the {@link SslContext} of this {@link VirtualHost}.
+     */
+    @SuppressWarnings("unchecked")
```

Comment:
To avoid repetitive `@SuppressWarnings` annotation, could you introduce the `self()` method?

```java
@SuppressWarnings("unchecked")
final B self() {
    return (B) this;
}
```